File: wallet.py
import ecdsa
from hashlib import sha256


# A wallet is a public-private keypair.
# The public key is the address.
# The private key is used to sign transactions.
# SECP256k1 is the Bitcoin elliptic curve used in this wallet.
class Wallet:
    def __init__(self, prvkey):
        self.prvkey = prvkey
    
    def __repr__(self) -> str:
        return self.pubkey_str()

    def pubkey(self):
        return self.prvkey.get_verifying_key()
    
    def pubkey_str(self):
        return self.pubkey().to_string().hex()
    
    def prvkey_str(self):
        return self.prvkey.to_string().hex()
    
    @staticmethod
    def create_random():
        prvkey = ecdsa.SigningKey.generate(curve=ecdsa.SECP256k1, hashfunc=sha256)
        return Wallet(prvkey)
    
    @staticmethod
    def from_private_key(private_key_hex):
        prvkey = ecdsa.SigningKey.from_string(bytes.fromhex(private_key_hex), curve=ecdsa.SECP256k1, hashfunc=sha256)
        return Wallet(prvkey)

    def sign(self, data):
        # assert type of `data` is bytes
        if type(data) != bytes:
            raise Exception("data must be bytes")
        return self.prvkey.sign(data)

    @staticmethod
    def verify(pubkey_hex, sig, msg):
        # assert type of `data` is bytes
        if type(msg) != bytes:
            raise Exception("data must be bytes")
        vk = ecdsa.VerifyingKey.from_string(bytes.fromhex(pubkey_hex), curve=ecdsa.SECP256k1, hashfunc=sha256)
        return vk.verify(sig, msg)


if __name__ == "__main__":
    w = Wallet.create_random()
    print("Wallet details:")
    print("  Private key: {}".format(w.prvkey_str()))
    print("  Public key: {}".format(w.pubkey_str()))

    msg = b"hello world"
    sig = w.sign(msg)
    print("Signature: {}".format(sig.hex()))

    is_valid = Wallet.verify(w.pubkey_str(), sig, msg)
    print("Verify signature: {}".format(is_valid))

File: core/wallet.go
package core

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"math/big"
	"fmt"
)

type Wallet struct {
	prvkey *ecdsa.PrivateKey
}

func (w *Wallet) Pubkey() *ecdsa.PublicKey {
	return &w.prvkey.PublicKey
}

func (w *Wallet) PubkeyBytes() [65]byte {
	pubkey := w.Pubkey()
	
	// 	The length of the buffer returned by elliptic.Marshal depends on the elliptic curve used. For the NIST P-256 curve (also known as elliptic.P256()), the buffer will be 65 bytes long. This includes:

	// 1 byte for the format prefix (0x04 for uncompressed)
	// 32 bytes for the X coordinate
	// 32 bytes for the Y coordinate

	buf := elliptic.Marshal(pubkey.Curve, pubkey.X, pubkey.Y)
	var pubkeyBytes [65]byte
	copy(pubkeyBytes[:], buf)
	return pubkeyBytes
}

func (w *Wallet) PubkeyStr() string {
	pubkey := w.PubkeyBytes()
	return hex.EncodeToString(pubkey[:])
}

func (w *Wallet) PrvkeyStr() string {
	return hex.EncodeToString(w.prvkey.D.Bytes())
}

func (w *Wallet) Address() string {
	pubkeyStr := w.PubkeyStr()
	firstHash := sha256.Sum256([]byte(pubkeyStr))
	secondHash := sha256.Sum256(firstHash[:])
	return hex.EncodeToString(secondHash[:])
}

func CreateRandomWallet() (*Wallet, error) {
	prvkey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		return nil, err
	}
	return &Wallet{prvkey: prvkey}, nil
}

func WalletFromPrivateKey(privateKeyHex string) (*Wallet, error) {
	privateKeyBytes, err := hex.DecodeString(privateKeyHex)
	if err != nil {
		return nil, err
	}
	prvkey := new(ecdsa.PrivateKey)
	prvkey.D = new(big.Int).SetBytes(privateKeyBytes)
	prvkey.PublicKey.Curve = elliptic.P256()
	prvkey.PublicKey.X, prvkey.PublicKey.Y = prvkey.PublicKey.Curve.ScalarBaseMult(privateKeyBytes)
	return &Wallet{prvkey: prvkey}, nil
}

func padBytes(src []byte, length int) []byte {
    if len(src) >= length {
        return src
    }
    padding := make([]byte, length-len(src))
    return append(padding, src...)
}

func (w *Wallet) Sign(msg []byte) ([]byte, error) {
	hash := sha256.Sum256(msg)
	r, s, err := ecdsa.Sign(rand.Reader, w.prvkey, hash[:])
	if err != nil {
		panic(err)
		return nil, err
	}
	// Ensure r and s are padded to 32 bytes
	rBytes := padBytes(r.Bytes(), 32)
	sBytes := padBytes(s.Bytes(), 32)

	// Concatenate r and s to form the signature
	signature := append(rBytes, sBytes...)
	
	return signature, nil
}

func VerifySignature(pubkeyStr string, sig, msg []byte) bool {
	if len(sig) != 64 {
		fmt.Printf("Invalid signature length: %s\n", len(sig)) // TODO
		return false
	}
	if len(pubkeyStr) != 130 {
		panic("Invalid public key") // TODO
		return false
	}

	pubkeyBytes, err := hex.DecodeString(pubkeyStr)
	if err != nil {
		panic(err)
		return false
	}

	x, y := elliptic.Unmarshal(elliptic.P256(), pubkeyBytes)
	if x == nil {
		panic("Invalid public key") // TODO
		return false
	}
	pubkey := &ecdsa.PublicKey{Curve: elliptic.P256(), X: x, Y: y}

	hash := sha256.Sum256(msg)
	r := new(big.Int).SetBytes(sig[:len(sig)/2])
	s := new(big.Int).SetBytes(sig[len(sig)/2:])

	// fmt.Println(r.Sign() <= 0 || s.Sign() <= 0)
	fmt.Printf("%s\n", pubkeyStr)
	fmt.Printf("r=%s s=%s\n", r.String(), s.String())

	return ecdsa.Verify(pubkey, hash[:], r, s)
}

